<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto City</font></h1>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [82]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

Using Power Query to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

* The data is filtered for te Not assigned neighbourhood (ignored)

In [83]:
URL='https://srv-file14.gofile.io/download/Fthpc5/Canada.csv'
df = pd.read_csv(URL)
df.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [84]:
df.shape
#df.dtypes

(103, 3)

In [85]:
link='http://cocl.us/Geospatial_data'
df2 = pd.read_csv(link)
df2.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [86]:
df2.shape
#df2.dtypes

(103, 3)

In [87]:
df  =df.set_index('Postal Code')
df.drop_duplicates(inplace=True)
df2  = df2.set_index('Postal Code')
df3  = df.join(df2,how='outer')
df3

Borough  \
Postal Code                     
M1B               Scarborough   
M1C               Scarborough   
M1E               Scarborough   
M1G               Scarborough   
M1H               Scarborough   
M1J               Scarborough   
M1K               Scarborough   
M1L               Scarborough   
M1M               Scarborough   
M1N               Scarborough   
M1P               Scarborough   
M1R               Scarborough   
M1S               Scarborough   
M1T               Scarborough   
M1V               Scarborough   
M1W               Scarborough   
M1X               Scarborough   
M2H                North York   
M2J                North York   
M2K                North York   
M2L                North York   
M2M                North York   
M2N                North York   
M2P                North York   
M2R                North York   
M3A                North York   
M3B                North York   
M3C                       NaN   
M3H                North York   
M3J                North York   
M3K                North York   
M3L                       NaN   
M3M                       NaN   
M3N                       NaN   
M4A                North York   
M4B                 East York   
M4C                 East York   
M4E              East Toronto   
M4G                 East York   
M4H                 East York   
M4J                 East York   
M4K              East Toronto   
M4L              East Toronto   
M4M              East Toronto   
M4N           Central Toronto   
M4P           Central Toronto   
M4R           Central Toronto   
M4S           Central Toronto   
M4T           Central Toronto   
M4V           Central Toronto   
M4W          Downtown Toronto   
M4X          Downtown Toronto   
M4Y          Downtown Toronto   
M5A          Downtown Toronto   
M5B          Downtown Toronto   
M5C          Downtown Toronto   
M5E          Downtown Toronto   
M5G          Downtown Toronto   
M5H          Downtown Toronto   
M5J          Downtown Toronto   
M5K          Downtown Toronto   
M5L          Downtown Toronto   
M5M                North York   
M5N           Central Toronto   
M5P           Central Toronto   
M5R           Central Toronto   
M5S          Downtown Toronto   
M5T          Downtown Toronto   
M5V          Downtown Toronto   
M5W          Downtown Toronto   
M5X          Downtown Toronto   
M6A                North York   
M6B                North York   
M6C                      York   
M6E                      York   
M6G          Downtown Toronto   
M6H              West Toronto   
M6J              West Toronto   
M6K              West Toronto   
M6L                North York   
M6M                      York   
M6N                      York   
M6P              West Toronto   
M6R              West Toronto   
M6S              West Toronto   
M7A          Downtown Toronto   
M7R               Mississauga   
M7Y              East Toronto   
M8V                 Etobicoke   
M8W                 Etobicoke   
M8X                 Etobicoke   
M8Y                 Etobicoke   
M8Z                 Etobicoke   
M9A                 Etobicoke   
M9B                 Etobicoke   
M9C                 Etobicoke   
M9L                North York   
M9M                North York   
M9N                      York   
M9P                 Etobicoke   
M9R                 Etobicoke   
M9V                 Etobicoke   
M9W                 Etobicoke   

                                                  Neighborhood   Latitude  \
Postal Code                                                                 
M1B                                             Malvern, Rouge  43.806686   
M1C                     Rouge Hill, Port Union, Highland Creek  43.784535   
M1E                          Guildwood, Morningside, West Hill  43.763573   
M1G                                                     Woburn  43.770992   
M1H                                                  Cedarbrae  43.773136   
M1J              

making sure that the dataset has all 1 boroughs and 103 neighborhoods.

In [88]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df3['Borough'].unique()),
        df3.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [89]:
toronto_data = df3[df3['Borough'] == 'Scarborough'].reset_index(drop=True)
toronto_data.head()

Borough                            Neighborhood   Latitude  Longitude
0  Scarborough                          Malvern, Rouge  43.806686 -79.194353
1  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711
3  Scarborough                                  Woburn  43.770992 -79.216917
4  Scarborough                               Cedarbrae  43.773136 -79.239476

In [90]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [91]:
# create map of Toronto using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, df3 in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Neighborhood']):
    label = '{}, {}'.format(df3, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [92]:
CLIENT_ID = '40TUM012AY1QSEBJISSYUMRL5KSHJTEUOQPEGWO2NCMFB0AM' # your Foursquare ID
CLIENT_SECRET = 'MYCH5UVVOJJNXRNRA4LPRFDFRKE1ENZF5DIJTCWRIQGJLUHG' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 500
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 40TUM012AY1QSEBJISSYUMRL5KSHJTEUOQPEGWO2NCMFB0AM
CLIENT_SECRET:MYCH5UVVOJJNXRNRA4LPRFDFRKE1ENZF5DIJTCWRIQGJLUHG


<a id='item2'></a>

## Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [93]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [94]:
# type your answer here

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )



Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge


Double-click __here__ for the solution.
<!-- The correct answer is:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )
--> 

#### Let's check the size of the resulting dataframe

In [95]:
print(toronto_venues.shape)
toronto_venues.head()

(84, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy’s       43.807448   
1              -79.160497  Royal Canadian Legion       43.782533   
2              -79.188711         RBC Royal Bank       43.766790   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711       Big Bite Burrito       43.766299   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.191151                  Bank  
3       -79.191537     Electronics Store  
4       -79.190720    Mexican Restaurant

Let's check how many venues were returned for each neighborhood

In [96]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Birch Cliff, Cliffside West                                             4   
Cedarbrae                                                               8   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Golden Mile, Clairlea, Oakridge                                         9   
Guildwood, Morningside, West Hill                                       7   
Kennedy Park, Ionview, East Birchmount Park                             6   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor...                      2   
Rouge Hill, Port Union, Highland Creek                                  1   
Scarborough Village                                                     2   
Steeles West, L'Amoreaux West                                          11   
Wexford, Maryvale                                                       4   
Woburn                                                                  5   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Agincourt                                                                5   
Birch Cliff, Cliffside West                                              4   
Cedarbrae                                                                8   
Clarks Corners, Tam O'Shanter, Sullivan                                 12   
Cliffside, Cliffcrest, Scarborough Village West                          2   
Dorset Park, Wexford Heights, Scarborough Town ...                       5   
Golden Mile, Clairlea, Oakridge                                          9   
Guildwood, Morningside, West Hill                                        7   
Kennedy Park, Ionview, East Birchmount Park                              6   
Malvern, Rouge                                                           1   
Milliken, Agincourt North, Steeles East, L'Amor...                       2   
Rouge Hill, Port Union, Highland Creek                                   1   
Scarborough Village                                                      2   
Steeles West, L'Amoreaux West                                           11   
Wexford, Maryvale                                                        4   
Woburn                                                                   5   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Agincourt                                               5               5   
Birch Cliff, Cliffside West                             4               4   
Cedarbrae                                               8               8   
Clarks Corners, Tam O'Shanter, Sullivan                12              12   
Cliffside, Cliffcrest, Scarborough Village West         2               2   
Dorset Park, Wexford Heights, Scarborough Town ...      5               5   
Golden Mile, Clairlea, Oakridge                         9               9   
Guildwood, Morningside, West Hill                       7               7   
Kennedy Park, Ionview, East Birchmount Park             6               6   
Malvern, Rouge                                          1               1   
Milliken, Agincourt North, Steeles East, L'Amor...      2               2   
Rouge Hill, Port Union, Highland Creek                  1               1   
Scarborough Village                                     2               2   
Steeles West, L'Amoreaux West                          11              11   
Wexford, Maryvale           

#### Let's find out how many unique categories can be curated from all the returned venues

In [97]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 49 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [98]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  American Restaurant  \
0                          Malvern, Rouge                    0   
1  Rouge Hill, Port Union, Highland Creek                    0   
2       Guildwood, Morningside, West Hill                    0   
3       Guildwood, Morningside, West Hill                    0   
4       Guildwood, Morningside, West Hill                    0   

   Athletics & Sports  Bakery  Bank  Bar  Breakfast Spot  Bus Line  \
0                   0       0     0    0               0         0   
1                   0       0     0    1               0         0   
2                   0       0     1    0               0         0   
3                   0       0     0    0               0         0   
4                   0       0     0    0               0         0   

   Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0            0     0                     0                   0   
1            0     0                     0                   0   
2            0     0                     0                   0   
3            0     0                     0                   0   
4            0     0                     0                   0   

   Clothing Store  Coffee Shop  College Stadium  Convenience Store  \
0               0            0                0                  0   
1               0            0                0                  0   
2               0            0                0                  0   
3               0            0                0                  0   
4               0            0                0                  0   

   Department Store  Discount Store  Electronics Store  Fast Food Restaurant  \
0                 0               0                  0                     1   
1                 0               0                  0                     0   
2                 0               0                  0                     0   
3                 0               0                  1                     0   
4                 0               0                  0                     0   

   Fried Chicken Joint  Gas Station  General Entertainment  Grocery Store  \
0                    0            0                      0              0   
1                    0            0                      0              0   
2                    0            0                      0              0   
3                    0            0                      0              0   
4                    0            0                      0              0   

   Hakka Restaurant  Ice Cream Shop  Indian Restaurant  Intersection  \
0                 0               0                  0             0   
1                 0               0                  0             0   
2                 0               0                  0             0   
3                 0               0                  0             0   
4                 0               0                  0             0   

   Italian Restaurant  Korean Restaurant  Latin American Restaurant  Lounge  \
0                   0                  0                          0       0   
1                   0                  0                          0       0   
2                   0                  0                          0       0   
3                   0                  0                          0       0   
4                   0                  0                          0       0   

   Medical Center  Mexican Restaurant  Middle Eastern Restaurant  Motel  \
0               0                   0                          0      0   
1               0                   0                          0      0   
2               0                   0                          0      0   
3               0                   0                          0      0   
4               0                   1                          0      0   

   Noodle House  Park  Pet Store  Pharmacy  Pizza Place  Playground  \
0             0    

And let's examine the new dataframe size.

In [99]:
toronto_onehot.shape

(84, 50)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [100]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  American Restaurant  \
0                                           Agincourt                  0.0   
1                         Birch Cliff, Cliffside West                  0.0   
2                                           Cedarbrae                  0.0   
3             Clarks Corners, Tam O'Shanter, Sullivan                  0.0   
4     Cliffside, Cliffcrest, Scarborough Village West                  0.5   
5   Dorset Park, Wexford Heights, Scarborough Town...                  0.0   
6                     Golden Mile, Clairlea, Oakridge                  0.0   
7                   Guildwood, Morningside, West Hill                  0.0   
8         Kennedy Park, Ionview, East Birchmount Park                  0.0   
9                                      Malvern, Rouge                  0.0   
10  Milliken, Agincourt North, Steeles East, L'Amo...                  0.0   
11             Rouge Hill, Port Union, Highland Creek                  0.0   
12                                Scarborough Village                  0.0   
13                      Steeles West, L'Amoreaux West                  0.0   
14                                  Wexford, Maryvale                  0.0   
15                                             Woburn                  0.0   

    Athletics & Sports    Bakery      Bank  Bar  Breakfast Spot  Bus Line  \
0                0.000  0.000000  0.000000  0.0        0.200000  0.000000   
1                0.000  0.000000  0.000000  0.0        0.000000  0.000000   
2                0.125  0.125000  0.125000  0.0        0.000000  0.000000   
3                0.000  0.000000  0.083333  0.0        0.000000  0.000000   
4                0.000  0.000000  0.000000  0.0        0.000000  0.000000   
5                0.000  0.000000  0.000000  0.0        0.000000  0.000000   
6                0.000  0.222222  0.000000  0.0        0.000000  0.222222   
7                0.000  0.000000  0.142857  0.0        0.142857  0.000000   
8                0.000  0.000000  0.000000  0.0        0.000000  0.000000   
9                0.000  0.000000  0.000000  0.0        0.000000  0.000000   
10               0.000  0.000000  0.000000  0.0        0.000000  0.000000   
11               0.000  0.000000  0.000000  1.0        0.000000  0.000000   
12               0.000  0.000000  0.000000  0.0        0.000000  0.000000   
13               0.000  0.000000  0.090909  0.0        0.090909  0.000000   
14               0.000  0.250000  0.000000  0.0        0.250000  0.000000   
15               0.000  0.000000  0.000000  0.0        0.000000  0.000000   

    Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0      0.000000  0.00                 0.000            0.000000   
1      0.000000  0.25                 0.000            0.000000   
2      0.000000  0.00                 0.125            0.000000   
3      0.000000  0.00                 0.000            0.083333   
4      0.000000  0.00                 0.000            0.000000   
5      0.000000  0.00                 0.000            0.200000   
6      0.111111  0.00                 0.000            0.000000   
7      0.000000  0.00                 0.000            0.000000   
8      0.166667  0.00                 0.000            0.000000   
9      0.000000  0.00                 0.000            0.000000   
10     0.000000  0.00                 0.000            0.000000   
11     0.000000  0.00                 0.000            0.000000   
12     0.000000  0.00                 0.000            0.000000   
13     0.000000  0.00                 0.000            0.181818   
14     0.000000  0.00                 0.000            0.000000   
15     0.000000  0.00                 0.000            0.000000   

    Clothing Store  Coffee Shop  College Stadium  Convenience Store  \
0              0.2     0.000000             0.00           0.000000   
1              0.0     0.000000             0.25           0.000000   
2              0.0     0.000000             0.00   

#### Let's confirm the new size

In [101]:
toronto_grouped.shape

(16, 50)

#### Let's print each neighborhood along with the top 5 most common venues

In [102]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1                     Lounge   0.2
2             Breakfast Spot   0.2
3               Skating Rink   0.2
4             Clothing Store   0.2


----Birch Cliff, Cliffside West----
                   venue  freq
0        College Stadium  0.25
1  General Entertainment  0.25
2           Skating Rink  0.25
3                   Café  0.25
4              Pet Store  0.00


----Cedarbrae----
                venue  freq
0              Bakery  0.12
1                Bank  0.12
2     Thai Restaurant  0.12
3  Athletics & Sports  0.12
4    Hakka Restaurant  0.12


----Clarks Corners, Tam O'Shanter, Sullivan----
                 venue  freq
0          Pizza Place  0.17
1    Convenience Store  0.08
2   Chinese Restaurant  0.08
3          Gas Station  0.08
4  Fried Chicken Joint  0.08


----Cliffside, Cliffcrest, Scarborough Village West----
                 venue  freq
0  American Restaurant   0.5
1                Mot

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [103]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [104]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  \
0                                        Agincourt         Clothing Store   
1                      Birch Cliff, Cliffside West  General Entertainment   
2                                        Cedarbrae        Thai Restaurant   
3          Clarks Corners, Tam O'Shanter, Sullivan            Pizza Place   
4  Cliffside, Cliffcrest, Scarborough Village West    American Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0          Skating Rink        Breakfast Spot  Latin American Restaurant   
1          Skating Rink                  Café            College Stadium   
2    Athletics & Sports                Bakery                       Bank   
3          Noodle House   Fried Chicken Joint            Thai Restaurant   
4                 Motel           Coffee Shop      General Entertainment   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Lounge           Coffee Shop           Gas Station   
1  Vietnamese Restaurant           Coffee Shop           Gas Station   
2            Gas Station   Fried Chicken Joint  Caribbean Restaurant   
3   Fast Food Restaurant    Italian Restaurant     Convenience Store   
4            Gas Station   Fried Chicken Joint  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Fried Chicken Joint  Fast Food Restaurant      Electronics Store  
1   Fried Chicken Joint  Fast Food Restaurant      Electronics Store  
2      Hakka Restaurant     Convenience Store          Grocery Store  
3    Chinese Restaurant              Pharmacy            Gas Station  
4     Electronics Store        Discount Store       Department Store

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [105]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 4, 1, 1, 1, 1, 1, 1, 1, 3])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [106]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Borough                            Neighborhood   Latitude  Longitude  \
0  Scarborough                          Malvern, Rouge  43.806686 -79.194353   
1  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711   
3  Scarborough                                  Woburn  43.770992 -79.216917   
4  Scarborough                               Cedarbrae  43.773136 -79.239476   

   Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0             3.0  Fast Food Restaurant  Vietnamese Restaurant   
1             0.0                   Bar  Vietnamese Restaurant   
2             1.0                  Bank           Intersection   
3             1.0           Coffee Shop           Soccer Field   
4             1.0       Thai Restaurant     Athletics & Sports   

  3rd Most Common Venue  4th Most Common Venue  5th Most Common Venue  \
0      Hakka Restaurant  General Entertainment            Gas Station   
1           Coffee Shop  General Entertainment            Gas Station   
2        Breakfast Spot    Rental Car Location      Electronics Store   
3     Korean Restaurant               Pharmacy  Vietnamese Restaurant   
4                Bakery                   Bank            Gas Station   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0   Fried Chicken Joint     Electronics Store         Discount Store   
1   Fried Chicken Joint  Fast Food Restaurant      Electronics Store   
2        Medical Center    Mexican Restaurant  Vietnamese Restaurant   
3        Clothing Store           Gas Station    Fried Chicken Joint   
4   Fried Chicken Joint  Caribbean Restaurant       Hakka Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store      Convenience Store  
1        Discount Store       Department Store  
2       College Stadium    Fried Chicken Joint  
3  Fast Food Restaurant      Electronics Store  
4     Convenience Store          Grocery Store

<a id='item5'></a>

<hr>

